pkg/ingestor/key/key.go

```diff
@@ -0,0 +1,213 @@
+//
+// Copyright 2022 The GUAC Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package key
+
+import (
+	"bytes"
+	"crypto"
+	"crypto/ecdsa"
+	"crypto/ed25519"
+	"crypto/rsa"
+	"crypto/sha256"
+	"encoding/hex"
+	"errors"
+	"fmt"
+
+	"github.com/sigstore/rekor/pkg/pki/x509"
+	"github.com/sigstore/sigstore/pkg/cryptoutils"
+	"github.com/sirupsen/logrus"
+)
+
+const (
+	rsaKeyType            string = "rsa"
+	ecdsaKeyType          string = "ecdsa"
+	ed25519KeyType        string = "ed25519"
+	rsassapsssha256Scheme string = "rsassa-pss-sha256"
+	ecdsaSha2nistp256     string = "ecdsa-sha2-nistp256"
+	ed25519Scheme         string = "ed25519"
+)
+
+type KeyProvider interface {
+	// RetrieveKey takes in the ID (which is commonly the hash of the key)
+	// and it retrieves the crypto.PublicKey that is associated
+	// Returns nil, nil if no keys are found
+	// Return nil, error if the request to the provider failed
+	RetrieveKey(id string) (crypto.PublicKey, error)
+	// StoreKey takes in the ID and the crypto.PublicKey and stores them
+	// for future retrieval. If key id is already present, it replaces
+	// the key with the new one
+	StoreKey(id string, pk crypto.PublicKey) error
+	// DeleteKey takes in the ID and will remove the associated key from the provider
+	DeleteKey(id string) error
+	// Type returns the key provider type
+	Type() KeyProviderType
+}
+
+type Key struct {
+	// KeyHash sha256 hash of the canonical representation of the key
+	KeyHash string
+	// KeyType represents the type of the key
+	KeyType string
+	// KeyVal is the crypto.PublicKey of the public key
+	KeyVal crypto.PublicKey
+	// TODO: is this needed? Santiago question?
+	// Scheme is the supported scheme by the key type.
+	Scheme string
+}
+
+type KeyProviderType string
+
+var (
+	keyProviders = map[KeyProviderType]KeyProvider{}
+)
+
+func RegisterKeyProvider(k KeyProvider, providerType KeyProviderType) {
+	if _, ok := keyProviders[providerType]; ok {
+		logrus.Warnf("the key provider is being overwritten: %s", providerType)
+	}
+	keyProviders[providerType] = k
+}
+
+// Find goes through each of the registered key providers and retrieves the key
+// TODO: Should this handle if multiple keys are returned
+func Find(id string) (*Key, error) {
```

Comment:
Find and Retrieve have quite a bit of re-used code, can we refactor for find to use retrieve?

Reply:
ah yes good catch.